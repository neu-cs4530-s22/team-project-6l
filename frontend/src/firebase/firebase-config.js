// Import the functions you need from the SDKs you need
import { initializeApp } from "firebase/app";
import { getAuth } from "firebase/auth";
// TODO: Add SDKs for Firebase products that you want to use
// https://firebase.google.com/docs/web/setup#available-libraries

// Your web app's Firebase configuration
// For Firebase JS SDK v7.20.0 and later, measurementId is optional
const firebaseConfig = {
  apiKey: "AIzaSyCY6FF-e9h48NmhLu3WrHdWOzn7bjN_i1g",
  authDomain: "coveytown-ac1e3.firebaseapp.com",
  projectId: "coveytown-ac1e3",
  storageBucket: "coveytown-ac1e3.appspot.com",
  messagingSenderId: "987672035225",
  appId: "1:987672035225:web:094db237029f4c824fdd08",
  measurementId: "G-6E85LP2QW8"
};

// Initialize Firebase
const app = initializeApp(firebaseConfig);
const auth = getAuth(app);
export default auth;