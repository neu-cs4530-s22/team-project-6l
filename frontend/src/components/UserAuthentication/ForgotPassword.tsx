import React from 'react';
import {
  Box,
  FormControl,
  FormLabel,
  Input,
  Button,
  Text,
  Flex,
  Alert,
  AlertIcon,
  AlertTitle,
  CloseButton,
} from '@chakra-ui/react';
import { useHistory } from 'react-router-dom';
import { sendPasswordResetEmail } from 'firebase/auth';
import auth from '../../firebaseAuth/firebase-config';
import authCheck from './authCheck';

/**
 * Display forgot password form, 
 * which will send a reset link to email if the users forget their password
 * and return error message if email is yet registered or invalid.
 */
export default function ForgotPassword():JSX.Element {
  const history = useHistory();
  const [email, setEmail] = React.useState('');
  const [isSent, setSent] = React.useState(false);
  const [isAlert, setAlert] = React.useState(false);
  const [alertMess, setAlertMess] = React.useState('');

  const onSignInClick = ((event: React.MouseEvent) => {
    event.preventDefault();
    history.push("/");
  });

  const onSendingClick = async (event: React.MouseEvent) => {
    event.preventDefault();
    await sendPasswordResetEmail(auth, email.trim())
    .then(() => setSent(true))
    .catch(error => {
      const {code} = error;
      setAlert(true);
      setAlertMess(authCheck(code));
    })
  }

  return (
    <Flex height="100vh" width="full" justifyContent="center" align="center" flexDirection="column">
      <Box
        textAlign="left"
        w="800px" maxW='lg'
        backgroundColor="white"
        borderRadius='lg'
        paddingTop="6"
        paddingBottom="6"
        paddingLeft="8"
        paddingRight="8"
        shadow="0 0 0 2px rgba(0, 0, 0, 0.2)"
      >
        {isSent ?
          <>
            <Text data-testid="sent-message" fontSize='sm' fontWeight="semibold">A reset link has been sent to your email.</Text>
            <Button data-testid="signin-btn" mt={4} type="submit" backgroundColor="blue.500" color="white" onClick={e => onSignInClick(e)}>Sign in</Button>
          </> :
          <>
            <Text fontSize='2xl' fontWeight="semibold" marginTop="2">Forgot Your Password?</Text>
            <Box textAlign="left" marginTop="2">
              <FormControl>
                <FormLabel>Enter your email</FormLabel>
                <Input data-testid="forgot-email" type="email" placeholder='Email' onChange={(event) => setEmail(event.target.value)} />
              </FormControl>
            </Box>
            {isAlert ?
            <Box marginTop="2">
              <Alert data-testid="error" status='error'>
                <AlertIcon />
                <AlertTitle data-testid="error-message" mr={2}>{alertMess}</AlertTitle>
                <CloseButton marginLeft="1" position='absolute' right='8px' top='8px' onClick={() => setAlert(false)} />
              </Alert>
            </Box> : <></>}
            <Box justifyContent="flex-end" display="flex">
              <Button mt={4} type="submit" backgroundColor="blue.500" color="white" onClick={e => onSendingClick(e)}> Reset </Button>
            </Box>
          </>
        }
      </Box>
    </Flex>
  );
}