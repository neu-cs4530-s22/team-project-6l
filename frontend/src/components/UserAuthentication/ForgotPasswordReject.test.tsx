import React from 'react';
import '@testing-library/jest-dom';
import { fireEvent, queryByTestId, render, waitFor } from '@testing-library/react';
import { sendPasswordResetEmail } from 'firebase/auth';
import ForgotPassword from './ForgotPassword';

const mockHistory = jest.fn();

jest.mock('../../firebaseAuth/firebase-config', () => ({
  auth: jest.fn().mockReturnThis(),
}));

jest.mock('firebase/auth', () => ({
  sendPasswordResetEmail: jest.fn().mockRejectedValue(() => Promise.reject()),
}));

describe('ForgotPassword', () => {
  afterEach(() => {
    jest.clearAllMocks();
    mockHistory.mockReset();
  });
  it('should succesfully send the email and return back to log0in page', async () => {
    const { getByTestId, getByText, container } = render(<ForgotPassword />);
    const email = getByTestId('forgot-email');
    fireEvent.change(email, { target: { value: 'testing@gmail.com' } });
    const btn = getByText('Reset');
    fireEvent.click(btn);
    await waitFor(() => {
      expect(sendPasswordResetEmail).toBeCalled();
    })
    // the snapshot should show the error message
    expect(container.firstChild).toMatchSnapshot();
    expect(queryByTestId(container, 'error-message')).toBeDefined();
  });
})