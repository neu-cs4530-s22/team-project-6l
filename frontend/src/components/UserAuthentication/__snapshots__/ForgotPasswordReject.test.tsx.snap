// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForgotPassword should succesfully send the email and return back to log0in page 1`] = `
<div
  class="css-1t8zqkj"
>
  <div
    class="css-15sedz4"
  >
    <p
      class="chakra-text css-13gk006"
    >
      Forgot Your Password?
    </p>
    <div
      class="css-ebxf5s"
    >
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-uvho5g"
          for="field-1"
          id="field-1-label"
        >
          Enter your email
        </label>
        <input
          class="chakra-input css-0"
          data-testid="forgot-email"
          id="field-1"
          placeholder="Email"
          type="email"
        />
      </div>
    </div>
    <div
      class="css-g89h0y"
    >
      <div
        class="chakra-alert css-1hwd9xd"
        data-testid="error-message"
        role="alert"
      >
        <span
          class="chakra-alert__icon css-1c6vscd"
        >
          <svg
            class="chakra-icon css-fi0ww0"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11.983,0a12.206,12.206,0,0,0-8.51,3.653A11.8,11.8,0,0,0,0,12.207,11.779,11.779,0,0,0,11.8,24h.214A12.111,12.111,0,0,0,24,11.791h0A11.766,11.766,0,0,0,11.983,0ZM10.5,16.542a1.476,1.476,0,0,1,1.449-1.53h.027a1.527,1.527,0,0,1,1.523,1.47,1.475,1.475,0,0,1-1.449,1.53h-.027A1.529,1.529,0,0,1,10.5,16.542ZM11,12.5v-6a1,1,0,0,1,2,0v6a1,1,0,1,1-2,0Z"
              fill="currentColor"
            />
          </svg>
        </span>
        <div
          class="chakra-alert__title css-1qkvt8e"
        >
          Error!!!
        </div>
        <button
          aria-label="Close"
          class="css-1mduyzg"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="chakra-icon css-onkibi"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="css-1958zak"
    >
      <button
        class="chakra-button css-1fibvam"
        type="submit"
      >
         Reset 
      </button>
    </div>
  </div>
</div>
`;
